### PR TITLE
Refactor test_sync_mode to use MockDevice

### PR DESCRIPTION
**Pull request by millerdev**
Probably best reviewed by commit 🐠 Don't miss the collapsed `test_sync_mode.py` file, which is where the bulk of the work happened, if reviewing all changes at once.

To assist in this refactoring effort, I created utilities (not part of this PR) to verify that...
- the XML posted to HQ by `MockDevice` is equivalent to what was posted before when interacting directly with the `CaseFactory`.
- the restore config and payload generated by `MockDevice.sync()` are equivalent to what was restored before when using the various restore helper/assert functions.

While these utilities were not used to refactor every single test, they gave me a degree of confidence that the refactoring has not changed what is being tested.

@proteusvacuum @snopoke cc @orangejenny

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 1]
```diff
@@ -2165,98 +1984,33 @@ class LiveQuerySteadyStateExtensionSyncTestSQL(LiveQuerySteadyStateExtensionSync
...
-            user_id=self.user_id,
-            owner_id=self.user_id,
-            case_type=PARENT_TYPE,
-        ).as_xml()
-        try:
-            self._postFakeWithSyncToken(caseblock, self.sync_log.get_id)
-            self.fail('posting an update to a non-existant case should fail')
-        except AssertionError:
-            # this should fail because it's a true error
-            pass
```
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: Removed this test because `self.fail(...)` raises `AssertionError`, which means the test will never fail. I have no idea what the "true error" comment was supposed to mean.

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 2]
```diff
@@ -2165,98 +1984,33 @@ class LiveQuerySteadyStateExtensionSyncTestSQL(LiveQuerySteadyStateExtensionSync
...
-
-        try:
-            post_case_blocks(
-                _get_bad_caseblocks([case_id2, case_id1]),
-                form_extras={ "last_sync_token": self.sync_log._id }
-            )
-            self.fail('posting an update to non-existant cases should fail')
-        except AssertionError:
-            # this should fail because it's a true error
-            pass
```
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: Removed this test because `self.fail(...)` raises `AssertionError`, which means the test will never fail. I have no idea what the "true error" comment was supposed to mean.

**Comment by millerdev** [comment]
@dannyroberts, I'd be grateful if you were able to review this.

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 3]
```diff
@@ -90,7 +90,7 @@ def setUp(self):
             self.restore_options,
             default_case_type=PARENT_TYPE,
         )
-        self.sync_log = self.device.last_sync.log
+        self.sync_log = self.device.sync(overwrite_cache=True, version=V1).log
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: What's the idea behind this?

File: corehq/ex-submodules/casexml/apps/phone/tests/utils.py [hunk 4]
```diff
@@ -171,7 +172,8 @@ def __init__(self, project, user, restore_options, default_case_type="case"):
             },
         )
         self.last_sync = None
-        self.sync(overwrite_cache=True)
+        if sync:
+            self.sync()
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/phone/tests/utils.py]: Why no `overwrite_cache=True` here?

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 5]
```diff
@@ -1977,8 +1977,8 @@ def create_case_graph(num):
         p2, o2, e2 = create_case_graph(2)
 
         # Alice and Bob sync
-        alice = self.get_device()
-        bob = self.get_device(user=self.other_user)
+        alice = self.get_device(sync=True)
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: Why isn't this just

```python
alice = self.get_device()
alice.sync()
```

I guess it's more lines. But it is more explicit about what's happening.

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 6]
```diff
@@ -165,7 +156,23 @@ def _testUpdate(self, sync_log_or_id, case_id_map, dependent_case_id_map=None):
             self._testUpdate(migrated_sync_log, case_id_map, dependent_case_id_map)
 
 
-class SyncTokenUpdateTest(SyncBaseTest):
+class SyncBaseTest(BaseSyncTest):
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: Maybe change this name to `DeprecatedBaseSyncTest`? (Right now the only difference in their names is transposing Sync and Base, which doesn't really tell you anything about which is preferred.)

File: corehq/ex-submodules/casexml/apps/case/tests/util.py [hunk 7]
```diff
@@ -105,20 +105,24 @@ def check_xml_line_by_line(test_case, expected, actual):
 
 
 def assert_user_has_case(testcase, user, case_id, **kwargs):
+    """DEPRECATED use assertIn(case_id, <MockDevice>.sync().cases)"""
```
Comment (dannyroberts) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: worth using

```python
warnings.warn("use assertIn(case_id, <MockDevice>.sync().cases)", DeprecationWarning)
```

or changing name to `deprecated_assert_user_has_case` instead?

**Review by dannyroberts — APPROVED** [review]
Looks great!

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py (hunk 3 above)
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: This is what it did [before](https://github.com/dimagi/commcare-hq/blob/aa13efd674f130c22b0f0e7ac769832684c524f5/corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py#L80-L83). `RestoreConfig` uses [`version=V1` by default](https://github.com/dimagi/commcare-hq/blob/5bd6c3ba709f380d9f14f6643e73ccb044fe8bef/corehq/ex-submodules/casexml/apps/phone/restore.py#L410).

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py (hunk 5 above)
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: it's just for convenience.

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py (hunk 6 above)
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: Good idea. I'll do it in a separate PR. It's used in a lot of places, which is why I did not change it in this PR.

File: corehq/ex-submodules/casexml/apps/phone/tests/utils.py (hunk 4 above)
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/tests/utils.py]: I don't think it's necessary. Does a sync without a sync token hit the cache?

File: corehq/ex-submodules/casexml/apps/case/tests/util.py (hunk 7 above)
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/case/tests/util.py]: Good idea. Will do this in a separate PR.
